CONTRIBUTING.md

```diff
@@ -0,0 +1,76 @@
+# Contributing
+
+## Overview
+
+This document explains the processes and practices recommended for contributing enhancements to
+this operator.
+
+- Generally, before developing enhancements to this charm, you should consider [opening an issue
+  ](https://github.com/canonical/training-operator/issues) explaining your use case.
+- If you would like to chat with us about your use-cases or proposed implementation, you can reach
+  us at [Canonical Mattermost public channel](https://chat.charmhub.io/charmhub/channels/charm-dev)
+  or [Discourse](https://discourse.charmhub.io/).
+- Familiarising yourself with the [Charmed Operator Framework](https://juju.is/docs/sdk) library
+  will help you a lot when working on new features or bug fixes.
+- All enhancements require review before being merged. Code review typically examines
+  - code quality
+  - test coverage
+  - user experience for Juju administrators of this charm.
+- Please help us out in ensuring easy to review branches by rebasing your pull request branch onto
+  the `main` branch. This also avoids merge commits and creates a linear Git commit history.
+
+## Developing
+
+You can use the environments created by `tox` for development:
+
+```shell
+tox --notest -e unit
+source .tox/unit/bin/activate
+```
+
+### Testing
+
+```shell
+tox -e lint          # code style
+tox -e unit          # unit tests
+tox -e integration   # integration tests
+tox                  # runs 'lint' and 'unit' environments
+```
+
+## Build charm
+
+Build the charm in this git repository using:
+
+```shell
+charmcraft pack
+```
+
+### Deploy
+
+```bash
+# Create a model
+juju add-model dev
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm
+juju deploy ./training-operator_ubuntu-20.04-amd64.charm \
+    --resource training-operator-image=kubeflow/training-operator
+```
+
+## Charmed Training Operator source and resources upgrade
```

Comment:
I generally like this, but if we're applying it to all of our charms maybe we can make it a bit more generic without losing any content.  There's a lot of mention of "Charmed Training Operator" and similar things that I think are unnecessary - the reader will know that's the operator since they're in this repo.  I'm thinking something like:

--------------------------

## Updating the charm for new versions of the workload
( not sure I like the above wording...  Previous wording felt very specific, but I think this is the procedure we got through whenever we bump the version of upstream that we're using.  It generally happens to bump resources and change source, but neither **must** occur.  Suggestions appreciated )

To update the workload charmed by this operator, you must:

1) Bump all `oci-image` resources in `metadata.yaml` to use the new desired version
2) Update the charm for any changes, such as:
  * YAML manifests in `src/` or `KubernetesResources` in pod_spec
  * new or changed configurations passed to pebble workloads or through `pod.set_spec`
3) Ensure integration and unit tests are passing; fix/adapt them otherwise

...

Reply:
* The tile: yeh, I wanted to be very specific because I didn't want to confuse the concept of upgrading the workload with upgrading the charm itself. I like your proposed title.
* Instructions: agree on 1, for 2 I would change it slightly:
    2. Update the charm source for any changes, such as:

    * YAML manifests in `src/` and/or any Kubernetes resource in pod_spec
    * new or changed configurations passed to pebble workloads or through pod.set_spec

Reply:
lgtm!

Reply:
+1 lets update it.